README.md

<p align="center">
  <img width=20% src="https://www.sap.com/content/dam/application/imagelibrary/pictograms/283000/283370-pictogram-purple.svg" alt="Logo" />
  </br>
  <span><b>SAP Conversational AI SDK for iOS</b></span>
</p>

***

<div align="center">
    <a href="https://github.com/SAP/cloud-sdk-ios-cai#installation">Installation
    </a>
    |
    <a href="https://github.com/SAP/cloud-sdk-ios-cai#examples"> Examples
    </a>
    |
    <a href="https://sap.github.io/cloud-sdk-ios-cai"> API Documentation
    </a>
    |
    <a href="https://github.com/SAP/cloud-sdk-ios-cai/blob/main/CHANGELOG.md"> Changelog
    </a>
</div>

***

<div align="center">
    <a href="https://github.com/SAP/cloud-sdk-ios-cai/actions?query=workflow%3A%22CI%22">
        <img src="https://github.com/SAP/cloud-sdk-ios-cai/workflows/CI/badge.svg"
             alt="Build Status">
    </a>
    <a href="https://conventionalcommits.org">
        <img src="https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg"
             alt="Conventional Commits">
    </a>
    <a href="http://commitizen.github.io/cz-cli/">
        <img src="https://img.shields.io/badge/commitizen-friendly-brightgreen.svg"
             alt="Commitizen friendly">
    </a>
    <a href="https://api.reuse.software/info/github.com/SAP/cloud-sdk-ios-cai">
        <img src="https://api.reuse.software/badge/github.com/SAP/cloud-sdk-ios-cai"
             alt="REUSE status">
    </a>
</div>

***

The `SAPCAI` Swift Package is a module to connect to a bot on the end-to-end chatbot platform [SAP Conversational AI](https://www.sap.com/products/conversational-ai.html) (CAI). It provides a pluggable SwiftUI `AssistantView` to initiate a conversation and render the various [message types](https://help.sap.com/viewer/a4522a393d2b4643812b7caadfe90c18/latest/en-US/ad3681adae824f8a96cbcf8b889a4ffc.html) of a bot.

This Swift Package is an open-source addition to the [SAP BTP SDK for iOS](https://developers.sap.com/topics/sap-btp-sdk-for-ios.html).

<p align="center">
<img src="https://user-images.githubusercontent.com/4176826/118739439-558fdd80-b7fe-11eb-9db2-032e210c5746.gif" alt="SwiftUI AssistantView" width="300" height="650" align="center">
</p>

## Requirements

- iOS 13 or higher
- Xcode 12 (Swift 5.3) or higher

## Prerequisites

- You have existing bot(s) on the SAP Conversational AI platform
- You created an OAuth client for Designtime API for your bot (<kbd>Settings</kbd> > <kbd> Tokens</kbd>)
- Your bot(s) are connected to a mobile channel

### Create a mobile channel

In the *Enterprise Edition* you create a mobile channel via the UI.

<img width="1093" alt="CreateMobileChannel" src="https://user-images.githubusercontent.com/4176826/119203198-549dbc80-ba47-11eb-8b65-d813324c3cc8.png">

<img alt="CreateMobileChannelResults" src="https://user-images.githubusercontent.com/4176826/119582931-6a2e2180-bd7a-11eb-8faf-1b801a03d146.png">

However, the option is currently hidden by default and must be enabled by filing a [ticket](http://help.sap.com/disclaimer?site=https%3A%2F%2Flaunchpad.support.sap.com%2F%23incident%2Fsolution).

In the *Community Edition* you create a mobile channel via the [/channel](https://reverseproxy.cai.tools.sap/docs/api-reference/#channels) API.
 
```bash
# create a mobile channel
curl -X POST "<BaseUrl>/connect/v1/channels" \
-H "X-Token: Token <DeveloperToken>" \
-H "Content-Type: application/json" \
-d '{"type":"mobile","slug":"<desiredChannelSlug>","connectionTarget": "SAP_Product","targetSystem": "MobileBot","isActivated":true}'
```

## Dependencies

- [SAPCommon](https://help.sap.com/doc/978e4f6c968c4cc5a30f9d324aa4b1d7/Latest/en-US/Documents/Frameworks/SAPCommon/index.html) for Logging
- [SAPFoundation](https://help.sap.com/doc/978e4f6c968c4cc5a30f9d324aa4b1d7/Latest/en-US/Documents/Frameworks/SAPFoundation/index.html) for Network Connectivity and Authentication
- [URLImage](https://github.com/dmytro-anokhin/url-image) for asynchronous image loading in SwiftUI
- [Down](https://github.com/johnxnguyen/Down) for Markdown / CommonMark rendering in Swift

## Installation

The package is intended for consumption via Swift Package Manager. To add `SAPCAI` to your application target, navigate to the  <kbd>File </kbd> >  <kbd>Swift Packages </kbd> >  <kbd>Add Package Dependency... </kbd>, then add the repository URL https://github.com/SAP/cloud-sdk-ios-cai.

You can choose one of the following package products to be added to your application/framework target.

|Package Product|When to Use
|---|---|
|SAPCAI|You did not already embed binary frameworks from SAP BTP SDK for iOS|
|SAPCAI-requiresToEmbedXCFrameworks|You already embedded `SAPCommon` and `SAPFoundation` binary frameworks to your target|

## Getting Started

The code snippet below illustrates setting up the `CAIConversation` Combine publisher to initialize the `MessagingViewModel` for the `AssistantView`. For prototyping you can connect the iOS client directly with SAP Conversational AI *Community Edition*. You have to replace variable values in angle brackets (e.g. `<channelId>`) with your bot-specific values.

See [Enterprise Configuration](https://github.com/SAP/cloud-sdk-ios-cai/blob/main/ENTERPRISE_CONFIG.md) guide for required configuration steps to use `SAPCAI` Swift package and connect indirectly with SAP Conversational AI *Enterprise Edition* through SAP Mobile Services.

### Setting up the data publisher

```Swift
import SAPCAI
import SAPFoundation
import SwiftUI

/*
  Create the `CAIChannel` object by providing the channel id, channel token, and channel slug, which were created in CAI Platform.

  If you do not want to hardcode channel information on the client, you can retrieve a list of channels for a given Application ID with `CAIChannelService`.
*/
let caiChannel = CAIChannel(id: "<channelId>", token: "<channelToken>", slug: "<channelSlug>").

// Create CAI service config for a SAPURLSession and URL pointing to CAI API endpoint
let serviceConfig = CAIServiceConfig(urlSession: SAPURLSession(), host: URL(string:"<Example: https://api.cai.tools.sap>")!)
     
// Provide the message delivery object for polling
let polling: MessageDelivering = PollMessageDelivery(channelId: "<channelId>", serviceConfig: serviceConfig)
   
// Create CAIConversation object and use it as publisher
let dataPublisher = CAIConversation(config: serviceConfig, channel: caiChannel, messageDelivery: polling)

// Create view model for `AssistantView`
let viewModel = MessagingViewModel(publisher: dataPublisher)
```

### User Interface

Using the SwiftUI implementation is the **preferred** approach.

```Swift
AssistantView()
  .navigationBarTitle(Text("Chat"), displayMode: .inline) // if you are in navigation controller
  .environmentObject(viewModel)
  .environmentObject(ThemeManager.shared)
  .onDisappear {
    viewModel.cancelSubscriptions()
    dataPublisher.resetConversation()
  })
```

but `SAPCAI` also provides a UIKit wrapper

```Swift
let vc = MessagingViewController(MessagingViewModel(publisher: dataPublisher))

self.navigationController?.pushViewController(vc, animated: true)
```

## Theming

`SAPCAI` will provide two standard themes
- a Fiori theme and
- a Casual theme

<img width="663" alt="Themes" src="https://user-images.githubusercontent.com/4176826/119032300-e1217f80-b960-11eb-8d65-e068958aff36.png">

Each theme will come with a color palette that supports both light and dark modes.

### Fiori theme

```swift
ThemeManager.shared.setCurrentTheme( .fiori(FioriColorPalette()) )
```

### Casual theme
    
```swift
ThemeManager.shared.setCurrentTheme( .casual(CasualColorPalette()) )
```

### Create your own custom theme

You can also provide an alternative color palette or provide a custom theme.

## Persistent Static Menu

In SAP Conversational AI you can add persistent static menu options to your chatbot enabling your users to quickly trigger links or skills at any point during the conversation.

<img width="525" alt="Persistent Static Menu In Bot Builder" src="https://user-images.githubusercontent.com/4176826/146964912-6cceec20-ea57-4bda-90f1-b6fa339eb470.png">

Use `CAIChannelService.loadPreferences` to fetch your channel's preferences (incl. the persistent static menu) asynchronously, then set `MessagingViewModel.menu`.

```Swift
service.loadPreferences(channel: channel) { result in
  DispatchQueue.main.async {
    switch result {
      case .success(let data):
        viewModel.menu = data?.menu
      case .failure:
        break
    }
  }
}
```

![Persistent Static Menu Rendered by SAP CAI SDK for iOS](https://user-images.githubusercontent.com/4176826/146965541-0288dfe5-99bf-47e9-9779-d196583d6253.png)


## Limitations

### Markdown related

- **No support for [GitHub Flavored Markdown](https://github.github.com/gfm/)** (e.g. ~~strikethrough~~). Bot responses containing [Markdown](https://help.sap.com/viewer/a4522a393d2b4643812b7caadfe90c18/latest/en-US/3a3b6e09e96249b884dcf6e69eecd830.html#loio3a3b6e09e96249b884dcf6e69eecd830__section_markdown) will be rendered by the `AssistantView` according to the [Commonmark specification](https://spec.commonmark.org/0.29/) only, and unknown syntax will be treated as text. 

### Trigger Skill button related

- `Trigger Skill` buttons are still enabled after execution / next interaction. This behavior differs from SAP CAI WebClient. Once you click the button or type an utterance, the corresponding skill can't be triggered anymore at a later point in time since the context in which it was created might not be valid anymore.

### Image related

- Not being able to save, copy, or share an image as part of a bot response. Currently, there is no gesture handler attached to an image view which could allow opening a contextual menu offering such features (similar to iMessage or WhatsApp).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.2](https://github.com/SAP/cloud-sdk-ios-cai/compare/2.0.1...2.0.2) (2022-03-11)

### Bug Fixes

* 🐛 Card Details: crash if clicking on button ([576e929](https://github.com/SAP/cloud-sdk-ios-cai/commit/576e929e06c0362bb4050931ec3b6132842c20f7))
* 🐛 fix runtime exception in ObjectMesssageView ([f0565f6](https://github.com/SAP/cloud-sdk-ios-cai/commit/f0565f69605fec223faf97301c1af3e718f8b5f8))

## [2.0.1](https://github.com/SAP/cloud-sdk-ios-cai/compare/2.0.0...2.0.1) (2021-12-21)

### Features

* 🎸 localized texts in 2 new language ([433324c](https://github.com/SAP/cloud-sdk-ios-cai/commit/433324cad8e175b2629cb7390e7a09c9c7bcd066)), closes [#58](https://github.com/SAP/cloud-sdk-ios-cai/issues/58)

### Bug Fixes

* 🐛 decoding error when loading channel preferences/menu ([8d8853b](https://github.com/SAP/cloud-sdk-ios-cai/commit/8d8853b8ff65e3e945ff0ac67011c19ae900b2e2)), closes [#64](https://github.com/SAP/cloud-sdk-ios-cai/issues/64)
* 🐛 postback/link value was shown in persistent static menu ([04f7881](https://github.com/SAP/cloud-sdk-ios-cai/commit/04f788125fd50864e6bd7d904dc432a96186c05c)), closes [#66](https://github.com/SAP/cloud-sdk-ios-cai/issues/66)
---

### [1.0.5](https://github.com/SAP/cloud-sdk-ios-cai/compare/1.0.4...1.0.5) (2021-12-21)

### Bug Fixes

* 🐛 decoding error when loading channel preferences/menu ([f9c5fd8](https://github.com/SAP/cloud-sdk-ios-cai/commit/f9c5fd8fe3bbf50ce27e0595fccc2295604d0162)), closes [#64](https://github.com/SAP/cloud-sdk-ios-cai/issues/64)
* 🐛 postback/link value was shown in persistent static menu ([987d3a4](https://github.com/SAP/cloud-sdk-ios-cai/commit/987d3a400886845258bbe882417dc8dcb5059001)), closes [#66](https://github.com/SAP/cloud-sdk-ios-cai/issues/66)

## [2.0.0](https://github.com/SAP/cloud-sdk-ios-cai/compare/1.0.4...2.0.0) (2021-11-10)

### ⚠ BREAKING CHANGES

* 🧨 new minimum deployment target is iOS 14 (previously: iOS 13)
* 🧨 [SDWebImage](https://github.com/SDWebImage/SDWebImage) replaces [URLImage](https://github.com/dmytro-anokhin/url-image) as package dependency 
* 🧨 no cleanup logic in `onDisappear` !! 

  For previous release(s) it was suggested to implement cleanup logic in `onDisappear` but you cannot rely that `onDisappear` will be called only once. The `AssistantView` might trigger further navigations causing the `AssistantView` to disappear and re-appear. Therefore cleanup logic should be handled elsewhere.
* 🧨 `NavigationView` is required in view hierarchy.

### Features

* 🎸 CardPageView to be used in Carousel and List ([a585c84](https://github.com/SAP/cloud-sdk-ios-cai/commit/a585c84cc16125263424f7b4fbc7d373d986ca6c))
* 🎸 carousel image: support sap-icon ([a1ef423](https://github.com/SAP/cloud-sdk-ios-cai/commit/a1ef423dcd6bd7b020809ec2e9567e8e3e2a13d5))
* 🎸 expandable plain long text messages with "View more" button ([#42](https://github.com/SAP/cloud-sdk-ios-cai/issues/42)) ([7437316](https://github.com/SAP/cloud-sdk-ios-cai/commit/7437316fee74f8bee11a9fbe1a902dc0f27b026c))
* 🎸 gif supported through SDWebImage ([#32](https://github.com/SAP/cloud-sdk-ios-cai/issues/32)) ([8280053](https://github.com/SAP/cloud-sdk-ios-cai/commit/8280053136dc4bebfd3f0aaa2009feace77c9bcf))


### Bug Fixes

* 🐛 resolve cycle reference for conversation operation ([3310416](https://github.com/SAP/cloud-sdk-ios-cai/commit/331041683d504fa4a29792e2bd8581f51f113ebf))
* 🐛 UI improvement for spacing and heights ([#33](https://github.com/SAP/cloud-sdk-ios-cai/issues/33)) ([aa7ff09](https://github.com/SAP/cloud-sdk-ios-cai/commit/aa7ff09212a07e1c60cc7f7012e9968485141774))

## [1.0.4](https://github.com/SAP/cloud-sdk-ios-cai/compare/1.0.3...1.0.4) (2021-08-09)

### Features

* 🎸 client-side handling of delayed messages ([2d3bed2](https://github.com/SAP/cloud-sdk-ios-cai/commit/2d3bed279e16c2184de224fef2cd99b057abab1c))
* 🎸 delayed message(s) are visually announced ([5015477](https://github.com/SAP/cloud-sdk-ios-cai/commit/5015477c39c54a03f1180b7291f773eede1fb2c2))


### Bug Fixes

* 🐛 "Trigger Skill" button inside card posts data correctly ([da9a309](https://github.com/SAP/cloud-sdk-ios-cai/commit/da9a3098048a38707287050b3041d50da525a16b))
* 🐛 action button layout for different size class ([#27](https://github.com/SAP/cloud-sdk-ios-cai/issues/27)) ([399f3b2](https://github.com/SAP/cloud-sdk-ios-cai/commit/399f3b29addb72b4a07bc84628c61aefc8fb8fcf))
* 🐛 avoid "A server error occurred" msgs related to polling ([942096a](https://github.com/SAP/cloud-sdk-ios-cai/commit/942096ae5832cfea3c2229846010a480cfe496d6))

CAITestApp/CAITestApp/HomeView.swift
import FioriThemeManager
import SAPCAI
import SwiftUI

struct HomeView: View {
    enum TestDataType: Int, RawRepresentable, CaseIterable {
        case `static` = 0
        case live = 1
    }

    enum ViewRepresentation: Int, RawRepresentable {
        case fullscreen = 0
        case modal = 1
    }

    enum ChannelSelection: Int, RawRepresentable {
        case appId = 0
        case token = 1
    }

    enum ConversationType: Int, RawRepresentable {
        case new = 0
        case existing = 1
    }

    @State private var selectedTestDataType: TestDataType = .static

    @State private var selectedTheme = 0

    @State private var selectedConnectionIndex: Int = 0 {
        didSet {
            if self.clients.count > self.selectedConnectionIndex {
                self.liveConnection.update(basedOn: self.clients[self.selectedConnectionIndex])
            }
        }
    }

    @State private var selectedViewRepresentation: ViewRepresentation = .fullscreen

    @State private var selectedChannelSelection: ChannelSelection = .appId

    @State private var isPresented = false

    @ObservedObject private var liveConnection = Connection()

    var clients: [Connection] = []

    let dataModel: DataModel

    init(dataModel: DataModel) {
        self.dataModel = dataModel
        self.selectedConnectionIndex = 0

        if let url = Bundle.main.url(forResource: "Connections", withExtension: "plist") {
            if let data = try? Data(contentsOf: url) {
                self.clients = try! PropertyListDecoder().decode([Connection].self, from: data)
            }
            self.clients.forEach { client in
                if let savedConnection = Connection.loadFromUserDefaults(with: client.name) {
                    client.update(basedOn: savedConnection)
                }
            }
        }

        // UINavigationBar.appearance().largeTitleTextAttributes = [.font: UIFont.boldSystemFont(ofSize: 28)]

        // largeTitle with 72 fonts is
        UINavigationBar.appearance().largeTitleTextAttributes = [.foregroundColor: UIColor.preferredFioriColor(forStyle: .primaryLabel), .font: UIFont.preferredFioriFont(forTextStyle: .largeTitle, weight: .black)]
    }

    var body: some View {
        NavigationView {
            ScrollView {
                VStack(alignment: .center, spacing: 15) {
                    // NavigationLink("Choose Colors", destination: ThemingView())

                    VStack(alignment: .leading, spacing: 4) {
                        Text("Theme")
                        ThemePickerView(selectedTheme: $selectedTheme)
                    }

                    VStack(alignment: .leading, spacing: 4) {
                        Section(header: Text("Representation")) {
                            Picker(selection: Binding<ViewRepresentation>(get: {
                                self.selectedViewRepresentation
                            }, set: { value in
                                self.selectedViewRepresentation = value
                                if value == .modal {
                                    self.selectedTestDataType = .static
                                }
                            }), label: Text("Representation")) {
                                Text("Fullscreen").tag(ViewRepresentation.fullscreen)
                                Text("Modal").tag(ViewRepresentation.modal)
                            }.pickerStyle(SegmentedPickerStyle())
                        }
                    }

                    VStack(alignment: .leading, spacing: 4) {
                        Text("Data")
                        Picker(selection: Binding<TestDataType>(get: {
                            self.selectedTestDataType
                        }, set: { value in
                            self.selectedTestDataType = value
                            if value == .live {
                                self.selectedViewRepresentation = .fullscreen
                                self.selectedConnectionIndex = 0
                            }
                        }), label: Text("Data")) {
                            Text("Static").tag(TestDataType.static)
                            if self.selectedViewRepresentation == .fullscreen {
                                Text("Live").tag(TestDataType.live)
                            }
                        }.pickerStyle(SegmentedPickerStyle())
                    }

                    if self.selectedTestDataType == TestDataType.live {
                        VStack(alignment: .leading, spacing: 4) {
                            Text("Connection")
                            Picker(selection: Binding<Int>(get: {
                                self.selectedConnectionIndex
                            }, set: { value in
                                self.selectedConnectionIndex = value
                                self.liveConnection.update(basedOn: clients[self.selectedConnectionIndex])
                            }), label: Text("Connection")) {
                                ForEach(0 ..< self.clients.count) { i in
                                    Text(self.clients[i].name).tag(i)
                                }
                            }.pickerStyle(SegmentedPickerStyle())

                            if clients[self.selectedConnectionIndex].name == "Custom" {
                                EditableConnectionView(model: liveConnection)
                            }

                            Text("Channel")
                            Picker(selection: Binding<ChannelSelection>(get: {
                                self.selectedChannelSelection
                            }, set: { value in
                                self.selectedChannelSelection = value
                            }), label: Text("Channel")) {
                                Text("Via AppId").tag(ChannelSelection.appId)
                                Text("Explicit").tag(ChannelSelection.token)
                            }.pickerStyle(SegmentedPickerStyle())

                            if self.selectedChannelSelection == ChannelSelection.appId {
                                Text("*Note: Channel Search via AppId does not work with SAP Conversational AI Community Edition. Use \"Explict\" in such case*. ")
                            }

                            if self.selectedChannelSelection == ChannelSelection.token {
                                VStack {
                                    FloatingTextField(title: "Channel ID", text: $liveConnection.channelId, placeholder: "Channel ID")
                                    FloatingTextField(title: "Channel Token", text: $liveConnection.channelToken, placeholder: "Channel Token")
                                    FloatingTextField(title: "Channel Slug", text: $liveConnection.channelSlug, placeholder: "Channel Slug")
                                    FloatingTextField(title: "Existing Conversation Id", text: $liveConnection.conversationId, placeholder: "Existing Conversation Id")

                                }.padding()
                            }
                        }
                    }

                    connectButton

                    Spacer()

                    Image("SAPLogo")
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .frame(height: 32, alignment: .center)
                }
                .padding(10)
                .navigationBarTitle("SAPCAI SDK for iOS")
                // .font(Font.fiori(forTextStyle: .largeTitle))
                .onAppear {
                    self.dataModel.clear()
                }
            }
        }
        .sheet(isPresented: self.$isPresented, onDismiss: {
            self.isPresented = false
        }, content: {
            NavigationView {
                AssistantView().navigationBarTitle(Text("Chat"), displayMode: .inline) // TODO: .large
                    .navigationBarItems(trailing: Button("Done", action: {
                        self.isPresented = false
                    }))
                    .environmentObject(MockData.viewModel)
                    .environmentObject(ThemeManager.shared)
            }.navigationViewStyle(StackNavigationViewStyle())

        })
        .navigationViewStyle(StackNavigationViewStyle())
    }

    @ViewBuilder var connectButton: some View {
        if self.selectedTestDataType == TestDataType.live {
            if self.selectedChannelSelection == ChannelSelection.token {
                NavigationLink(destination: NavigationLazyView(AssistantView()
                        .navigationBarTitle(Text(self.liveConnection.channelSlug), displayMode: .inline)
                        .environmentObject(self.dataModel.getModelFor(self.liveConnection))
                        .environmentObject(ThemeManager.shared)
                )) {
                    Text("Connect")
                }
            } else {
                connectButtonForChannelList()
            }
        } else {
            if self.selectedViewRepresentation == ViewRepresentation.fullscreen {
                NavigationLink(destination:
                    NavigationLazyView(
                        AssistantView().navigationBarTitle(Text("Chat"), displayMode: .inline) // TODO: large
                            .environmentObject(MockData.viewModel)
                            .environmentObject(ThemeManager.shared)
                    )
                ) {
                    Text("Connect")
                }
            }
            if self.selectedViewRepresentation == ViewRepresentation.modal {
                Button("Connect") {
                    self.isPresented = true
                }
            }
        }
    }

    func connectButtonForChannelList() -> AnyView {
        do {
            let config = try CAIServiceConfig.create(from: self.liveConnection)

            return AnyView(NavigationLink(destination:
                NavigationLazyView(
                    ChannelListView(CAIChannelService(config: config))
                        .environmentObject(self.dataModel)
                        .environmentObject(ThemeManager.shared))
            ) {
                Text("Connect")
            })
        } catch {
            return AnyView(Text("Enter correct URLs").foregroundColor(.red))
        }
    }
}

struct HomeView_Previews: PreviewProvider {
    static var previews: some View {
        HomeView(dataModel: DataModel())
    }
}

struct NavigationLazyView<Content: View>: View {
    let build: () -> Content
    init(_ build: @autoclosure @escaping () -> Content) {
        self.build = build
    }

    var body: Content {
        self.build()
    }
}

// ONLY FOR THIS TEST APP
// The following code is NOT needed in apps built with SAPFiori.xcframework as such methods are available there
// Introduced as `UINavigationBar.appearance().largeTitleTextAttributes` expects UIKit-based representations of color and font
private extension UIColor {
    static func preferredFioriColor(forStyle style: ColorStyle) -> UIColor {
        let color = Color.preferredColor(style)
        let uiColor = UIColor(color)
        return uiColor
    }
}

private extension UIFont {
    class func preferredFioriFont(forTextStyle textStyle: UIFont.TextStyle, weight: UIFont.Weight = .regular, isItalic: Bool = false, isCondensed: Bool = false) -> UIFont {
        guard var font = UIFont(name: get72FontName(weight: weight.fioriWeight), size: textStyle.size) else {
            var font = UIFont.preferredFont(forTextStyle: textStyle).withWeight(weight)

            if isItalic {
                font = font.italic
            }
//            if isCondensed {
//                font = font.with(.traitCondensed)
//            }

            return font
        }

        let metrics: UIFontMetrics
//        if textStyle == .KPI || textStyle == .largeKPI {
//            metrics = UIFontMetrics(forTextStyle: .largeTitle)
//        } else {
        metrics = UIFontMetrics(forTextStyle: textStyle)
//        }

        if isItalic {
            font = font.italic
        }
//        if isCondensed {
//            font = font.with(.traitCondensed)
//        }

        let scaledFont = metrics.scaledFont(for: font)

        return scaledFont
    }

    class func preferredFioriFont(fixedSize size: CGFloat, weight: UIFont.Weight = .regular, isItalic: Bool = false, isCondensed: Bool = false) -> UIFont {
        guard var font = UIFont(name: get72FontName(weight: weight.fioriWeight), size: size) else {
            var font = UIFont.systemFont(ofSize: size, weight: weight)

            if isItalic {
                font = font.italic
            }
            if isCondensed {
                font = font.with(.traitCondensed)
            }

            return font
        }

        if isItalic {
            font = font.italic
        }
        if isCondensed {
            font = font.with(.traitCondensed)
        }

        return font
    }

    static func get72FontName(weight: UIFont.Weight) -> String {
        let description = weight.description

        // TODO: waiting for designer to update font name in 72-black.ttf
        if weight == .black {
            return "72\(description)"
        }

        return "72-\(description)"
    }
}

private extension UIFont {
    var bold: UIFont {
        self.with(.traitBold)
    }

    var italic: UIFont {
        self.with(.traitItalic)
    }

    var boldItalic: UIFont {
        self.with([.traitBold, .traitItalic])
    }

    func with(_ traits: UIFontDescriptor.SymbolicTraits...) -> UIFont {
        guard let descriptor = self.fontDescriptor.withSymbolicTraits(UIFontDescriptor.SymbolicTraits(traits).union(self.fontDescriptor.symbolicTraits)) else {
            return self
        }
        return UIFont(descriptor: descriptor, size: 0)
    }

    func without(_ traits: UIFontDescriptor.SymbolicTraits...) -> UIFont {
        guard let descriptor = self.fontDescriptor.withSymbolicTraits(self.fontDescriptor.symbolicTraits.subtracting(UIFontDescriptor.SymbolicTraits(traits))) else {
            return self
        }
        return UIFont(descriptor: descriptor, size: 0)
    }

    func withWeight(_ weight: UIFont.Weight) -> UIFont {
        let newDescriptor = fontDescriptor.addingAttributes([.traits: [UIFontDescriptor.TraitKey.weight: weight]])
        return UIFont(descriptor: newDescriptor, size: 0)
    }
}

private extension UIFont.TextStyle {
    var size: CGFloat {
        switch self {
        case .largeTitle:
            return 34
        case .title1:
            return 28
        case .title2:
            return 22
        case .title3:
            return 20
        case .headline:
            return 17
        case .body:
            return 17
        case .callout:
            return 16
        case .subheadline:
            return 15
        case .footnote:
            return 13
        case .caption1:
            return 12
        case .caption2:
            return 11
//        case .largeKPI:
//            return 48
//        case .KPI:
//            return 36
        default:
            return 17
        }
    }
}

extension UIFont.TextStyle: CustomStringConvertible {
    /// :nodoc:
    public var description: String {
        switch self {
        case .largeTitle:
            return "largeTitle"
        case .title1:
            return "title1"
        case .title2:
            return "title2"
        case .title3:
            return "title3"
        case .headline:
            return "headline"
        case .body:
            return "body"
        case .callout:
            return "callout"
        case .subheadline:
            return "subheadline"
        case .footnote:
            return "footnote"
        case .caption1:
            return "caption1"
        case .caption2:
            return "caption2"
//        case .KPI:
//            return "KPI"
//        case .largeKPI:
//            return "largeKPI"
        default:
            return "Unknown"
        }
    }
}

extension UIFont.Weight: CaseIterable, CustomStringConvertible {
    /// :nodoc:
    public static var allCases: [UIFont.Weight] {
        [.black, .heavy, .bold, .semibold, .medium, .regular, .light, .thin, .ultraLight]
    }

    /// :nodoc:
    public var description: String {
        let name: String

        switch self {
        case .black:
            name = "black"
        case .heavy:
            name = "heavy"
        case .bold:
            name = "bold"
        case .semibold:
            name = "semibold"
        case .medium:
            name = "medium"
        case .regular:
            name = "regular"
        case .light:
            name = "light"
        case .thin:
            name = "thin"
        case .ultraLight:
            name = "ultraLight"
        default:
            name = "Unknown"
        }

        return name
    }
}

private extension UIFont.Weight {
    // Available 72 weights
    var fioriWeight: UIFont.Weight {
        switch self {
        case .heavy, .black:
            return .black
        case .medium, .semibold, .bold:
            return .bold
        case .regular:
            return .regular
        case .ultraLight, .thin, .light:
            return .light
        default:
            return .regular
        }
    }
}

private extension UIFontDescriptor {
    internal class func preferredFioriDescriptor(textStyle: String) -> UIFontDescriptor {
        enum Static {
            static var fontNameTable = [String: String]()
            static var fontSizeTable = [String: [String: Int]]()
            static var fontWeightTable = [String: String]()
        }

        enum KPINumber1Size {
            static let xsmall = Int(42)
            static let small = Int(44)
            static let medium = Int(46)
            static let large = Int(48)
            static let xlarge = Int(50)
            static let xxlarge = Int(52)
            static let xxxlarge = Int(54)
        }

        enum KPINumber2Size {
            static let xsmall = Int(29)
            static let small = Int(31)
            static let medium = Int(33)
            static let large = Int(35)
            static let xlarge = Int(37)
            static let xxlarge = Int(39)
            static let xxxlarge = Int(41)
        }

//        Static.fontSizeTable = [
//            FioriUIFontTextStyle.kpiNumber.rawValue: [UIContentSizeCategory.extraSmall.rawValue: KPINumber1Size.xsmall, UIContentSizeCategory.small.rawValue: KPINumber1Size.small, UIContentSizeCategory.medium.rawValue: KPINumber1Size.medium, UIContentSizeCategory.large.rawValue: KPINumber1Size.large, UIContentSizeCategory.extraLarge.rawValue: KPINumber1Size.xlarge, UIContentSizeCategory.extraExtraLarge.rawValue: KPINumber1Size.xxlarge, UIContentSizeCategory.extraExtraExtraLarge.rawValue: KPINumber1Size.xxxlarge, UIContentSizeCategory.accessibilityMedium.rawValue: KPINumber1Size.xxxlarge, UIContentSizeCategory.accessibilityLarge.rawValue: KPINumber1Size.xxxlarge, UIContentSizeCategory.accessibilityExtraLarge.rawValue: KPINumber1Size.xxxlarge, UIContentSizeCategory.accessibilityExtraExtraLarge.rawValue: KPINumber1Size.xxxlarge, UIContentSizeCategory.accessibilityExtraExtraExtraLarge.rawValue: KPINumber1Size.xxxlarge],
//
//            FioriUIFontTextStyle.kpiNumber2.rawValue: [UIContentSizeCategory.extraSmall.rawValue: KPINumber2Size.xsmall, UIContentSizeCategory.small.rawValue: KPINumber2Size.small, UIContentSizeCategory.medium.rawValue: KPINumber2Size.medium, UIContentSizeCategory.large.rawValue: KPINumber2Size.large, UIContentSizeCategory.extraLarge.rawValue: KPINumber2Size.xlarge, UIContentSizeCategory.extraExtraLarge.rawValue: KPINumber2Size.xxlarge, UIContentSizeCategory.extraExtraExtraLarge.rawValue: KPINumber2Size.xxxlarge, UIContentSizeCategory.accessibilityMedium.rawValue: KPINumber2Size.xxxlarge, UIContentSizeCategory.accessibilityLarge.rawValue: KPINumber2Size.xxxlarge, UIContentSizeCategory.accessibilityExtraLarge.rawValue: KPINumber2Size.xxxlarge, UIContentSizeCategory.accessibilityExtraExtraLarge.rawValue: KPINumber2Size.xxxlarge, UIContentSizeCategory.accessibilityExtraExtraExtraLarge.rawValue: KPINumber2Size.xxxlarge],
//        ]

//        Static.fontWeightTable = [
//            FioriUIFontTextStyle.kpiNumber.rawValue: "Thin",
//            FioriUIFontTextStyle.kpiNumber2.rawValue: "Light",
//        ]
//
//        Static.fontNameTable = [
//            FioriUIFontTextStyle.kpiNumber.rawValue: "SFUIDisplay",
//            FioriUIFontTextStyle.kpiNumber2.rawValue: "SFUIDisplay"
//        ]

        var contentSize = UIContentSizeCategory.large

        // This is a way to find out if UIApplication is initialized or not for framework unit testing
        // we could get EXC_BAD_ACCESS in calling UIApplication.shared().preferredContentSizeCategory when UIApplication object is not initialized yet
        if UIApplication.shared.description != "" {
            contentSize = UIApplication.shared.preferredContentSizeCategory
        }

        let style = Static.fontSizeTable[textStyle]!
        let fontName = "\(Static.fontNameTable[textStyle]!)-\(Static.fontWeightTable[textStyle]!)"

        return UIFontDescriptor(name: fontName, size: CGFloat(style[contentSize.rawValue]!))
    }
}
